### PR TITLE
[RFC-17] Add WithdrawalRequestCanceled Event to DepositManager

### PR DESCRIPTION
**Pull request by Zena-park**
## Related RFC
- **RFC Discussion**: https://github.com/tokamak-network/tokamak-dao-contracts/discussions/17
- **Original Proposal**: https://github.com/tokamak-network/tokamak-dao-contracts/discussions/16

## Summary

This PR implements RFC-17 by adding a `WithdrawalRequestCanceled` event to the `DepositManager` contract. This enables accurate tracking of TON circulating supply by distinguishing between "fresh deposits" and "withdrawal cancellations (redeposits)".

### Contract Changes
- ✅ Created `DepositManagerV1_2` contract with `WithdrawalRequestCanceled` event
- ✅ Modified `_redeposit` function to emit both `Deposited` and `WithdrawalRequestCanceled` events
- ✅ No breaking changes - only additive event emission

### Test Coverage
- ✅ Standalone tests: [`test/deposit-manager-v1-2-standalone.test.ts`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/test/deposit-manager-v1-2-standalone.test.ts)
- ✅ Agenda-based registration tests: [`test/deposit-manager-v1-2-agenda.test.ts`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/test/deposit-manager-v1-2-agenda.test.ts)
- ✅ Shared helper functions: [`test/shared/depositManagerHelpers.ts`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/test/shared/depositManagerHelpers.ts)
- ✅ All tests passing

## Implementation Details

When `redeposit` or `redepositMulti` is called, both events are emitted:
```solidity
emit Deposited(layer2, msg.sender, accAmount);
emit WithdrawalRequestCanceled(layer2, msg.sender, accAmount);
```

## Testing

All tests passing ✅
- Event emission verification
- E2E flow testing
- Agenda-based registration testing
- Edge cases (rounding, multiple withdrawals)

스크린샷 2025-12-05 오후 7 27 11
스크린샷 2025-12-05 오후 7 27 52

## Impact
- **Breaking Changes**: ❌ None 
- **Security**: No security impact (event-only change)

## References
- RFC document: [`docs/RFC/RFC-17-DepositManagerV1_2.md`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/docs/RFC/RFC-17-DepositManagerV1_2.md) 
- Contract: [`contracts/stake/managers/DepositManagerV1_2.sol`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/contracts/stake/managers/DepositManagerV1_2.sol)

### CODE REVIEW

File: contracts/stake/managers/DepositManagerV1_2.sol [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    function redepositMulti(address layer2, uint256 n) external returns (bool) {
+      uint256 i = _withdrawalRequestIndex[layer2][msg.sender];
+      require(_redeposit(layer2, i, n), "fail redeposit");
+      return true;
+    }
+
+    function _redeposit(address layer2, uint256 i, uint256 n) internal onlyLayer2(layer2) returns (bool) {
+      uint256 accAmount;
+
+      WithdrawalReqeust[] memory requsts = _withdrawalRequests[layer2][msg.sender];
```
[the remark this answers is not in the document]
Comment (Zena-park) [reply contracts/stake/managers/DepositManagerV1_2.sol]: Thank you! 
I've fixed the typo in the local variable: `requsts` → `requests`.

The `WithdrawalReqeust` type name is part of the storage layout, so I've left it unchanged to avoid compatibility issues. I'll fix it when we have a storage layout change in the future.

Fix commit: [5aa5eaf8](https://github.com/tokamak-network/ton-staking-v2/commit/5aa5eaf8)

File: contracts/stake/managers/DepositManagerV1_2.sol [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    DepositManagerStorage,
+    DepositManagerV1_1Storage
+{
+
+    ////////////////////
+    // Modifiers
+    ////////////////////
+
+    modifier onlyLayer2(address layer2) {
+      require(ILayer2Registry(_registry).layer2s(layer2), "Caller is not a Layer2");
```
[the remark this answers is not in the document]
Comment (Zena-park) [reply contracts/stake/managers/DepositManagerV1_2.sol]: Thank you! I've updated the error message to "Address is not a registered Layer2".

Fix commit: [5aa5eaf8](https://github.com/tokamak-network/ton-staking-v2/commit/5aa5eaf8)

File: contracts/stake/managers/DepositManagerV1_2.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        uint256 amount = r.amount;
+
+        require(!r.processed, "DepositManager: pending request already processed");
+        require(amount > 0, "DepositManager: no valid pending request");
+
+        accAmount = accAmount + amount;
+        r.processed = true;
+        _withdrawalRequests[layer2][msg.sender][i] = r;
+      }
```
[the remark this answers is not in the document]
Comment (Zena-park) [reply contracts/stake/managers/DepositManagerV1_2.sol]: Thank you! I've applied the gas optimization improvements.

- Removed array memory copy → read only length
- Used storage pointer → direct modification
- Removed storage write in loop

**Commit**: [21d21c9](https://github.com/tokamak-network/ton-staking-v2/pull/330/commits/21d21c96aef082c603d11dae1277e4db31229a37)

The array growth issue will be considered in future architectural improvements.

File: contracts/stake/managers/DepositManagerV1_2.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      _pendingUnstaked[layer2][msg.sender] = _pendingUnstaked[layer2][msg.sender] - accAmount;
+      _pendingUnstakedLayer2[layer2] = _pendingUnstakedLayer2[layer2] - accAmount;
+      _pendingUnstakedAccount[msg.sender] = _pendingUnstakedAccount[msg.sender] - accAmount;
+
+      _withdrawalRequestIndex[layer2][msg.sender] += n;
+
+      emit Deposited(layer2, msg.sender, accAmount);
+
+      // add event for withdrawal request canceled
+      emit WithdrawalRequestCanceled(layer2, msg.sender, accAmount);
```
Comment (shinthom) [inline contracts/stake/managers/DepositManagerV1_2.sol]: I’d like to suggest a small improvement for better readability:
- Place each event (emit) after its corresponding state variables are updated.
- Since the logic is essentially “cancel and then re-deposit,” it may be clearer to update the order accordingly as well (purely for readability).

For example:

```solidity
_withdrawalRequestIndex[layer2][msg.sender] += n;

// withdrawal-related storages
_pendingUnstaked[layer2][msg.sender] = _pendingUnstaked[layer2][msg.sender] - accAmount;
_pendingUnstakedLayer2[layer2] = _pendingUnstakedLayer2[layer2] - accAmount;
_pendingUnstakedAccount[msg.sender] = _pendingUnstakedAccount[msg.sender] - accAmount;
emit WithdrawalRequestCanceled(layer2, msg.sender, accAmount);

// deposit-related storages
_accStaked[layer2][msg.sender] = _accStaked[layer2][msg.sender] + accAmount;
_accStakedLayer2[layer2] = _accStakedLayer2[layer2] + accAmount;
_accStakedAccount[msg.sender] = _accStakedAccount[msg.sender] + accAmount;
emit Deposited(layer2, msg.sender, accAmount);
```

**Review by shinthom — APPROVED** [review]

File: contracts/stake/managers/DepositManagerV1_2.sol (hunk 4 above)
Comment (Zena-park) [reply contracts/stake/managers/DepositManagerV1_2.sol]: Thank you! I've reordered the code as suggested.

**Applied commit**: [4732e9c3](https://github.com/tokamak-network/ton-staking-v2/commit/4732e9c3)

**Comment by zzooppii** [comment]
Are you testing the version currently being tested on mainnet?

When creating an agenda, the method of creating an agenda differs depending on whether it is created before Agenda15 is passed or not, so please check only that part.

File: contracts/stake/managers/DepositManagerV1_2.sol (hunk 4 above)
Comment (Zena-park) [reply contracts/stake/managers/DepositManagerV1_2.sol]: @shinthom Does this change look good to you? If everything is clear, may I resolve this conversation?

**Comment by Zena-park** [comment]
> Are you testing the version currently being tested on mainnet?
> 
> When creating an agenda, the method of creating an agenda differs depending on whether it is created before Agenda15 is passed or not, so please check only that part.

@zzooppii I haven't simulated the actual agenda creation script yet.
I will write the simulation script first and let you know once it's ready. I would appreciate it if you could review it again at that time.

File: contracts/stake/managers/DepositManagerV1_2.sol (hunk 4 above)
Comment (shinthom) [reply contracts/stake/managers/DepositManagerV1_2.sol]: Sure, good to me!
